internal/oci/notation/notation.go

```diff
@@ -0,0 +1,380 @@
+		o.keychain = key
+	}
+}
+
+// WithLogger is a function that returns an Options function to set the logger for the options.
+// The logger is used for logging purposes within the options.
+func WithLogger(logger logr.Logger) Options {
+	return func(o *options) {
+		o.logger = logger
+	}
+}
+
+// NotationVerifier is a struct which is responsible for executing verification logic
+type NotationVerifier struct {
+	auth     authn.Authenticator
+	keychain authn.Keychain
+	verifier *notation.Verifier
+	opts     []remote.Option
+	insecure bool
+	logger   logr.Logger
+}
+
+var _ truststore.X509TrustStore = &trustStore{}
+
+// trustStore is used by notation-go/verifier to retrieve the root certificate for notary.
+// The default behaviour is to read the certificate from disk and return it as a byte slice.
+// The reason for implementing the interface here is to avoid reading the certificate from disk
+// as the certificate is already available in memory.
+type trustStore struct {
+	cert []byte
+}
+
+// GetCertificates implements truststore.X509TrustStore.
+func (s trustStore) GetCertificates(ctx context.Context, storeType truststore.Type, namedStore string) ([]*x509.Certificate, error) {
+	raw := s.cert
+	block, _ := pem.Decode(raw)
+	if block != nil {
+		raw = block.Bytes
+	}
+
+	certs, err := x509.ParseCertificates(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate '%s': %s", namedStore, err)
+	}
+
+	return certs, nil
+}
+
+// NewNotationVerifier initializes a new Verifier
+func NewNotationVerifier(opts ...Options) (*NotationVerifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	store := &trustStore{
+		cert: o.rootCertificate,
+	}
+
+	trustpolicy := cleanTrustPolicy(o.trustPolicy, o.logger)
+	if trustpolicy == nil {
+		return nil, fmt.Errorf("trust policy cannot be empty")
+	}
+
+	verifier, err := verifier.New(trustpolicy, store, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &NotationVerifier{
+		auth:     o.auth,
+		keychain: o.keychain,
+		verifier: &verifier,
+		opts:     o.rOpt,
+		insecure: o.insecure,
+		logger:   o.logger,
+	}, nil
+}
+
+// cleanTrustPolicy cleans the given trust policy by removing trust stores and trusted identities
+// for trust policy statements that are set to skip signature verification but still have configured trust stores and/or trusted identities.
+// It takes a pointer to a trustpolicy.Document and a logger from the logr package as input parameters.
+// If the trustPolicy is nil, it returns nil.
+// Otherwise, it iterates over the trustPolicy.TrustPolicies and checks if each trust policy statement's
+// SignatureVerification.VerificationLevel is set to trustpolicy.LevelSkip.Name.
+// If it is, it logs a warning message and removes the trust stores and trusted identities for that trust policy statement.
+// Finally, it returns the modified trustPolicy.
+func cleanTrustPolicy(trustPolicy *trustpolicy.Document, logger logr.Logger) *trustpolicy.Document {
```

Comment:
This function should probably be called in the reconciler so it can be called only once and not for every new `verifier`.

Reply:
Resolved in 99b8cb5